tests/bsim/bluetooth/host/l2cap/multilink_peripheral/src/dut.c

```diff
@@ -0,0 +1,337 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/l2cap.h>
+#include <zephyr/logging/log.h>
+
+#include "babblekit/flags.h"
+#include "babblekit/testcase.h"
+
+/* local includes */
+#include "data.h"
+
+LOG_MODULE_REGISTER(dut, CONFIG_APP_LOG_LEVEL);
+
+static DEFINE_FLAG(ADVERTISING);
+
+static void sdu_destroy(struct net_buf *buf)
+{
+	LOG_DBG("%p", buf);
+
+	net_buf_destroy(buf);
+}
+
+/* Only one SDU per link will be transmitted at a time */
+NET_BUF_POOL_DEFINE(sdu_tx_pool, CONFIG_BT_MAX_CONN, BT_L2CAP_SDU_BUF_SIZE(SDU_LEN),
+		    CONFIG_BT_CONN_TX_USER_DATA_SIZE, sdu_destroy);
+
+static uint8_t tx_data[SDU_LEN];
+
+struct test_ctx {
+	bt_addr_le_t peer;
+	struct bt_l2cap_le_chan le_chan;
+	size_t sdu_count; /* the number of SDUs that have been transferred until now */
+};
+
+static struct test_ctx contexts[CONFIG_BT_MAX_CONN];
+
+static int send_data_over_l2cap(struct bt_l2cap_chan *chan, uint8_t *data, size_t len)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(chan->conn), addr, sizeof(addr));
+
+	LOG_DBG("[%s] chan %p data %p len %d", addr, chan, data, len);
+
+	struct net_buf *buf = net_buf_alloc(&sdu_tx_pool, K_NO_WAIT);
+
+	if (buf == NULL) {
+		TEST_FAIL("No more memory");
+		return -ENOMEM;
+	}
+
+	net_buf_reserve(buf, BT_L2CAP_SDU_CHAN_SEND_RESERVE);
+	net_buf_add_mem(buf, data, len);
+
+	int ret = bt_l2cap_chan_send(chan, buf);
+
+	TEST_ASSERT(ret == 0, "Failed sending: err %d", ret);
+	LOG_DBG("queued SDU", len);
+
+	return ret;
+}
+
+static void continue_sending(struct test_ctx *ctx)
+{
+	struct bt_l2cap_chan *chan = &ctx->le_chan.chan;
+
+	TEST_ASSERT(ctx->le_chan.state == BT_L2CAP_CONNECTED, "such race much wow");
+
+	LOG_DBG("%p, transmitted %d SDUs", chan, ctx->sdu_count);
+
+	if (ctx->sdu_count < SDU_NUM) {
+		send_data_over_l2cap(chan, tx_data, sizeof(tx_data));
+	} else {
+		char addr[BT_ADDR_LE_STR_LEN];
+
+		bt_addr_le_to_str(bt_conn_get_dst(chan->conn), addr, sizeof(addr));
+
+		LOG_DBG("[%s] Done sending", addr);
+	}
+}
```

Comment:
Semantically it's weird that the function called `continue_sending` is responsible for determining if it should continue sending. At least to me, it makes much more sense that the call to `continue_sending` is guarded by `ctx->sdu_count < SDU_NUM` so that this function actually continues sending. 

The alternative would be to rename it `continue_sending_but_only_if_we_want_to_send_more_otherwise_just_log`

Reply:
I kinda like the long name :D